closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -380,30 +399,99 @@ JNIEXPORT jboolean JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_isOpenSSLFIP
     return OSSL_IS_FIPS;
 }
 
-/*
- * Class:     jdk_crypto_jniprovider_NativeCrypto
- * Method:    loadCrypto
- * Signature: (Z)J
- */
-JNIEXPORT jlong JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
-  (JNIEnv *env, jclass thisObj, jboolean trace)
+/* Unload the crypto library. */
+static void
+unload_crypto_library(void *crypto_library)
 {
-    typedef const char* OSSL_version_t(int);
+#if defined (_WIN32)
+    FreeLibrary(crypto_library);
+#else /* defined(_WIN32) */
+    (void)dlclose(crypto_library);
+#endif /* defined(_WIN32) */
+}
 
-    /* Determine the version of OpenSSL. */
-    OSSL_version_t* OSSL_version;
-    const char * openssl_version;
-    jlong ossl_ver = 0;
+/* Find the symbol in the crypto library (return NULL if not found). */
+static void *
+find_crypto_symbol(void *crypto_library, const char *symname)
+{
+    #if defined (_WIN32)
+    return GetProcAddress(crypto_library, symname);
+    #else
+    return dlsym(crypto_library, symname);
+    #endif
+}
 
-    /* Load OpenSSL Crypto library */
-    crypto_library = load_crypto_library(trace);
-    if (NULL == crypto_library) {
-        if (trace) {
-            fprintf(stderr, "Error loading OpenSSL: FAILED TO LOAD OPENSSL CRYPTO LIBRARY\n");
-            fflush(stderr);
+static void
+log_crypto_library_path(jboolean traceEnabled, void *crypto_library, char *message)
+{
+    if (traceEnabled && (NULL != crypto_library)) {
+
+#if defined(_AIX)
+        /* Initialize the buffer with maximum size for L_GETINFO. */
+        char *buffer = (char *)malloc(DLFCN_LDINFO_SIZE);
+        if (NULL == buffer) {
+            return;
         }
-        return -1;
+        int rc = -1;
+        /* Get the list of all object files loaded by this process. */
+        rc = loadquery(L_GETINFO, buffer, DLFCN_LDINFO_SIZE);
+
+        /* Parse the list of all object files and print the OPENSSL library path. */
+        if (0 == rc) {
+            char *buf = buffer;
+            struct ld_info* cur_info = NULL;
+            do {
+                cur_info = (struct ld_info *)buf;
+                char *path = cur_info->ldinfo_filename;
+                char *member_name = path + strlen(cur_info->ldinfo_filename) + 1;
+                if (('\0' != *member_name) && (NULL != strstr(path, "libcrypto"))) {
```

Comment:
I see what you mean, but I cannot think of a more appropriate pattern given the potential library names in AIX. This is also just for debug, so printing an unwanted library is not such a huge issue.

Reply:
Well, in AIX `libcrypto64.so` is also available. And we might even get it from an archive file, meaning that it would look something like `libcrypto.a(libcrypto.so)`. So, I'm not sure I see anything that covers all those. I could add the `/` though.

Reply:
Added.